sling/nlp/parser/trainer/transition-generator.cc

```diff
@@ -0,0 +1,457 @@
+// Copyright 2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <deque>
+#include <vector>
+
+#include "sling/frame/object.h"
+#include "sling/frame/store.h"
+#include "sling/nlp/parser/trainer/transition-generator.h" 
+
+namespace sling {
+namespace nlp {
+namespace {
+
+// Map from a frame to information pertinent to transition generation.
+class FrameInfo {
+ public:
+  // Represents one incoming/outgoing edge of a frame.
+  struct Edge {
+    Edge(Handle r, bool in, Handle n) : role(r), neighbor(n), incoming(in) {}
+
+    // Edge label.
+    Handle role;
+
+    // Other end of the edge. 
+    Handle neighbor;
+
+    // Whether the edge is coming into this frame.
+    bool incoming;
+      
+    // Pointer to the reverse edge. Not owned.
+    Edge *inverse = nullptr; 
+        
+    // Whether the edge has been used to output action(s).
+    bool accounted = false;
+  };    
+
+  // Information about a single frame.
+  struct Entry {
+    ~Entry() {
+      for (auto *edge : edges) delete edge;
+    }
+
+    Handle handle = Handle::nil();  // frame handle
+    Handle type = Handle::nil();    // type of the frame
+    bool span_based = false;        // whether the frame is evoked by text
```

Comment:
You could consider using "evoked" instead of "span_based", i.e. a frame that is evoked from a span in contrast to a frame that is not evoked from a frame.

Reply:
Done.